### PR TITLE
NonlinearEigenSystem: don't close the matrix before solve

### PR DESCRIPTION
**Pull request by fdkong**
If nonlinear eigenvalue solvers are chosen. The memory will be shrunk and the preallocaiton will be removed if we close the matrix too early. 

We need to wait for libMesh/libmesh#1459. 

Refs #7398

### CODE REVIEW

**Comment by fdkong** [comment]
@permcody, please take a look at this simple change.

File: framework/src/base/NonlinearEigenSystem.C [hunk 1]
```diff
@@ -119,6 +122,12 @@ NonlinearEigenSystem::solve()
   // Initialize the solution vector using a predictor and known values from nodal bcs
   setInitialSolution();
 
+// In DEBUG mode, Libmesh will check the residual automatically. This may cause
+// an error because B does not need to assembly by default.
+#ifdef DEBUG
```
Comment (permcody) [inline framework/src/base/NonlinearEigenSystem.C]: When you say debug mode, do you mean ONLY debug mode, or do you mean "not optimized" mode (i.e. also included DEVEL mode). Can you point me to the libMesh code in question? I can help answer this.

**Review by permcody — CHANGES_REQUESTED** [review]

File: framework/src/base/NonlinearEigenSystem.C (hunk 1 above)
Comment (fdkong) [reply framework/src/base/NonlinearEigenSystem.C]: I think I meant ONLY debug mode.

https://github.com/libMesh/libmesh/blob/7f3ced2710c67ef3f7d7ba512090259b09997c74/src/solvers/slepc_eigen_solver.C#L510

Comment (permcody) [reply framework/src/base/NonlinearEigenSystem.C]: Perfect! Thanks

**Review by permcody — APPROVED** [review]

**Pull request merged**